### PR TITLE
What's new sections for v4.0

### PR DESCRIPTION
**Pull request by astrofrog**
The goal of this PR is to add entries to the 'What's new in v4.0?'. I've added section headings for things I think are worth highlighting, but of course we could add/remove if needed. I suggest that we try and write all of these as part of a single PR as it will make it easier to go over after and make sure everything is consistent etc.

So if you would like to write one or more of these sections (or another section that is missing) please either just push a commit directly to my ``whatsnew-4.0`` branch or open a PR to my branch if you prefer or don't have the adequate permissions to push to my branch. If you aren't sure how to do either of these, feel free to ping me on slack.

I'm going to tag and assign people who've been involved in developing the functionality/changes in this page or oversaw the relevant PRs:

* [x] Cosmology: @aconley
* [x] New models: @karllark 
* [x] Table: @taldcroft 
* [x] Constants and units: @stargaser (draft done by @mhvk)
* [x] Galactocentric frame updates: @adrn 
* [x] New ymdhms Time format: ~@mhvk~ @taldcroft
* [x] Plotting Time values: @astrofrog
* [x] Parsing high-precision values in Time: @aarchiba (draft by @mhvk)
* [x] Improved compatibility of quantity with numpy ufuncs: @mhvk
* [x] 1-d profile plots and default labeling with WCSAxes: @Cadair 
* [x] Function to fit WCS to pairs of pixel/world coordinates: @eteq
* [x] WCS transformations with Time: @astrofrog 
* [x] Time series fold() improvements: @astrofrog 
* [x] Cache import/export: @aarchiba 
* [x] Performance improvements for table: @taldcroft 
* [x] Improved handling of leap seconds: @mhvk 
* [x] Modeling API changes: @nden
* [x] Uncertainty/Distribution API changes: @mhvk

If you don't have time to work on this over the next few days, please put a message and I'll find someone else who can write it or I can write it myself.

Tips:

* If you are a maintainer, please also check over https://github.com/astropy/astropy/pulls?q=is%3Apr+label%3Awhatsnew-needed+milestone%3Av4.0+is%3Aclosed to see if any other features/improvements for your sub-packages need to be mentioned and feel free to add a new entry.
* To avoid using up too much CI time, please be sure to include ``[docs only]`` in your commit messages!
* Don't force push to this branch!

Once most of the sections are here I'll go over and try and make sure everything is consistent.

### CODE REVIEW

**Comment by adrn** [comment]
Putting @lglattly on notice: We will want you to copy edit this PR once it is text-complete!

**Comment by adrn** [comment]
You should be able to rebase now that #9582 is merged, but I'm slightly disturbed that nothing is failing because of my addition to the whats new!

**Comment by astrofrog** [comment]
@adrn - I had added ``IGNORE_OUTPUT`` to some of the lines in your examples so I could test mine and forgot to then remove those flags. If I remove them then I get failures in this PR with the dict of references, and after rebasing those failures go away.

**Comment by lglattly** [comment]
> Putting @lglattly on notice: We will want you to copy edit this PR once it is text-complete!

Is this ready for a copy editing pass now?

**Comment by taldcroft** [comment]
I'm planning to add my sections this weekend.

**Comment by bsipocz** [comment]
> @lglattly - Is this ready for a copy editing pass now?

not yet, there are still significant points missing.

**Comment by astrofrog** [comment]
I am working on the missing sections now.

**Comment by astrofrog** [comment]
@aarchiba - would you be able to fill out the section on 'New cache import/export utilities'?

**Comment by karllark** [comment]
Apologies for not getting to my assigned sections.  Life busy.

**Comment by astrofrog** [comment]
I've cleaned up the page, and decided to get rid of the different levels of hierarchy that some sub-packages had and not others for consistency with other what's new pages and because then all the API changes can be included at the bottom.

**Comment by aarchiba** [comment]
I've made a PR with some description, but I wasn't sure of the format or how best to arrange for it to be included here.

**Comment by astrofrog** [comment]
@aarchiba - thanks! Your changes are now included here :)

@lglatty - this is now ready for copy editing, thanks!

**Comment by lglattly** [comment]
I made some copy editing suggestions that you can take or leave as you see fit!

File: docs/whatsnew/4.0.rst [hunk 1]
```diff
@@ -25,13 +41,579 @@ smaller improvements and bug fixes, which are described in the
...
+
+A key area was the handling of replacing and adding columns, which is now 2 to
+10 times faster in common cases. The implementation was changed so that the time
+for replacing or adding is independent of the number of existing columns (like a
+``dict``). This means one can now *efficiently* build a table from scratch by
+creating an empty table and then adding columns one at a time.
+
+Another improvement was in the performance of table and column slicing. In
+addition to internal implementation changes, there was a change to reduce
+unnecessary copy and deepcopy of table and column ``meta`` attributes. In
```
[the remark this answers is not in the document]
Comment (bsipocz) [reply docs/whatsnew/4.0.rst]: deepcopy here refers to the function, so I suppose it's better to keep its name

**Comment by bsipocz** [comment]
Hmm, it's quite annoying, I was clicking through all the copy edit comments, but at the end couldn't commit them due to permission error. Anyway, once they are in, this should be good to go.

File: docs/whatsnew/4.0.rst [hunk 2]
```diff
@@ -25,13 +41,579 @@ smaller improvements and bug fixes, which are described in the
...
+
+Improved consistency of physical constants and units
+====================================================
+
+Physical constants and units using them are based on measurements that improve
+over time and therefore are not "constant" numerically.  Generally, one wants to
+just use the latest values, but sometimes it is necessary to reproduce earlier
+results by going back in time.  For that purpose, we have now introduced new
+``astropy.physical_constants`` and ``astropy.astronomical_constants`` science
+state objects, which can be used to set to use previous versions, and to make
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply docs/whatsnew/4.0.rst]: The new wording didn't quite convey the meaning here so I reworded it in a separate commit.

File: docs/whatsnew/4.0.rst [hunk 3]
```diff
@@ -25,13 +41,579 @@ smaller improvements and bug fixes, which are described in the
...
+                     0.2
+
+.. _whatsnew-4.0-table-new:
+
+New Table methods and options
+=============================
+
+A new method `~astropy.table.dstack` was added to allow depth-wise stacking
+of tables to turn a list of similar tables into a single "3-d" table with
+(rows, columns, depth).
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply docs/whatsnew/4.0.rst]: I changed this to make it clear this is the shape of the resulting 3D table

File: docs/whatsnew/4.0.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (astrofrog) [reply docs/whatsnew/4.0.rst]: Agree, so I skipped this suggestion

File: docs/whatsnew/4.0.rst [hunk 4]
```diff
@@ -25,13 +41,579 @@ smaller improvements and bug fixes, which are described in the
...
+
+For numerical formats, :class:`~astropy.time.Time` can now be instantiated
+from strings, quadruple precision ``numpy`` floats (if available on a given
+platform), and :class:`~decimal.Decimal` instances. For instance, in the example below
+you can see how in a string we can give full precision, while entering the same
+number as a float gives precision loss::
+
+    >>> from astropy.time import Time
+    >>> t = Time('2450000.123456789012345', format='jd')
+    >>> t - Time(2450000, .123456789012345, format='jd')
```
Comment (taldcroft) [inline docs/whatsnew/4.0.rst]:
```suggestion
    >>> t - Time(2450000, 0.123456789012345, format='jd')
```

File: docs/whatsnew/4.0.rst [hunk 5]
```diff
@@ -25,13 +41,579 @@ smaller improvements and bug fixes, which are described in the
...
+  ``outputs`` of a model can be renamed.
+
+* Assigning slices of the model parameter array does not automatically get
+  reflected in parameter values
+
+* Previously it was possible to use arbitrary slices on compound models (which had
+  the possibility of returning submodes with entirely different meanings than they
+  had in the original compound model). Now only a restricted set of slices is permitted.
+
+* Use of “imputed” units is much more restricted. Previously these could end up with
```
Comment (pllim) [inline docs/whatsnew/4.0.rst]: Is this a typo?

```suggestion
* Use of “inputed” units is much more restricted. Previously these could end up with
```

**Comment by pllim** [comment]
A glance through the rendered doc didn't raise any red flags, except for the potential typo in above comment. Once that is resolved, this can be merged.

**Comment by bsipocz** [comment]
Thanks @astrofrog and everyone else involved!

**Pull request merged**